### PR TITLE
GuidanceViewImageProviderTest refactoring

### PR DESCRIPTION
**Pull request by korshaknn**
## Description

Simplify test's logic, remove unnecessary code.

We don't need a real callback object to handle `CountDownLatch`, we can use a mocked one with
`every { callback.onFailure(capture(messageSlot)) } answers { latch.countDown() }`

also `Slot` is suitable, no need to use Atomic refs 

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

## Testing

Please describe the manual tests that you ran to verify your changes

- [ ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [ ] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR
- [ ] We might need to update / push `api/current.txt` files after running `$> make core-update-api` (Core) / `$> make ui-update-api` (UI) if there are changes / errors we're 🆗 with (e.g. `AddedMethod` changes are marked as errors but don't break SemVer) 🚀 If there are SemVer breaking changes add the `SEMVER` label. See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs

### CODE REVIEW

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt [hunk 1]
```diff
@@ -70,18 +67,14 @@ class GuidanceViewImageProviderTest {
...
 
         guidanceViewImageProvider.renderGuidanceView(bannerInstructions, callback)
 
-        verify(exactly = 1) { callback.onFailure(capture(captureData)) }
-        assertEquals("Guidance View Image URL is null", captureData.captured)
+        verify(exactly = 1) { callback.onFailure(capture(messageSlot)) }
+        assertEquals(ERROR_VIEW_IMAGE_URL_NULL, messageSlot.captured)
     }
 
     @Test
```
Comment (Guardiola31337) [inline libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt:80]: `= coroutineRule.runBlockingTest` 👇 is not necessary

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProvider.kt [hunk 2]
```diff
@@ -29,6 +29,7 @@ class GuidanceViewImageProvider {
     companion object {
         private const val USER_AGENT_KEY = "User-Agent"
         private const val USER_AGENT_VALUE = "MapboxJava/"
+        internal const val ERROR_VIEW_IMAGE_URL_NULL = "Guidance View Image URL is null"
```
[the remark this answers is not in the document]
Comment (korshaknn) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProvider.kt:32]: done

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProvider.kt [hunk 3]
```diff
@@ -29,6 +29,7 @@ class GuidanceViewImageProvider {
     companion object {
```
[the remark this answers is not in the document]
Comment (korshaknn) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProvider.kt:29]: done

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt [hunk 4]
```diff
@@ -70,18 +67,14 @@ class GuidanceViewImageProviderTest {
...
+        val messageSlot = slot<String>()
+
+        every { bannerView.components() } returns mutableListOf(bannerComponentsWithNullGuidanceUrl())
 
         guidanceViewImageProvider.renderGuidanceView(bannerInstructions, callback)
 
-        verify(exactly = 1) { callback.onFailure(capture(captureData)) }
-        assertEquals("Guidance View Image URL is null", captureData.captured)
+        verify(exactly = 1) { callback.onFailure(capture(messageSlot)) }
+        assertEquals(ERROR_VIEW_IMAGE_URL_NULL, messageSlot.captured)
```
[the remark this answers is not in the document]
Comment (korshaknn) [reply libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt:77]: done

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt [hunk 5]
```diff
@@ -70,18 +67,14 @@ class GuidanceViewImageProviderTest {
...
     fun `when guidance view url is null`() {
-        val captureData = slot<String>()
-        val bannerInstructions: BannerInstructions = mockk()
-        val bannerView: BannerView = mockk()
-        val bannerComponentList =
-            mutableListOf<BannerComponents>(BannerComponentsFaker.bannerComponentsWithNullGuidanceUrl())
-        every { bannerInstructions.view() } returns bannerView
-        every { bannerView.components() } returns bannerComponentList
+        val messageSlot = slot<String>()
+
```
[the remark this answers is not in the document]
Comment (korshaknn) [reply libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt:71]: done

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt [hunk 6]
```diff
@@ -90,46 +83,30 @@ class GuidanceViewImageProviderTest {
...
-        val bannerView: BannerView = mockk()
-        val bannerComponentList =
-            mutableListOf<BannerComponents>(BannerComponentsFaker.bannerComponentsWithGuidanceUrlNoAccessToken(mockWebServer.url("guidance-views/v1/1580515200/jct/CA075101?arrow_ids=CA07510E&access_token=null").toString()))
-        every { bannerInstructions.view() } returns bannerView
-        every { bannerView.components() } returns bannerComponentList
+        every { bannerView.components() } returns mutableListOf(
+            bannerComponentsWithGuidanceUrlNoAccessToken(
+                mockWebServer.url(URL_WITH_NULL_TOKEN).toString()
+            )
+        )
```
[the remark this answers is not in the document]
Comment (korshaknn) [reply libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt:91]: done

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt [hunk 7]
```diff
@@ -90,46 +83,30 @@ class GuidanceViewImageProviderTest {
...
 
+        val messageSlot = slot<String>()
         val latch = CountDownLatch(1)
-        val bitmapRef = AtomicReference<Bitmap>()
-        val failureRef = AtomicBoolean()
-        val failureMessageRef = AtomicReference<String?>()
-        val aCallback = provideACallback(latch, bitmapRef, failureRef, failureMessageRef)
```
[the remark this answers is not in the document]
Comment (korshaknn) [reply libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt:94]: done

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt [hunk 8]
```diff
@@ -90,46 +83,30 @@ class GuidanceViewImageProviderTest {
         mockWebServer.enqueue(MockResponse().setResponseCode(401))
         mockWebServer.start()
```
[the remark this answers is not in the document]
Comment (korshaknn) [reply libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt:85]: done

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt [hunk 9]
```diff
@@ -90,46 +83,30 @@ class GuidanceViewImageProviderTest {
...
-        }
-        job.join()
+        every { callback.onFailure(capture(messageSlot)) } answers { latch.countDown() }
+
+        guidanceViewImageProvider.renderGuidanceView(bannerInstructions, callback)
 
         latch.await()
-        assertTrue(failureRef.get())
-        assertEquals("Client Error", failureMessageRef.get())
+
```
[the remark this answers is not in the document]
Comment (korshaknn) [reply libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt:100]: done

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt [hunk 10]
```diff
@@ -90,46 +83,30 @@ class GuidanceViewImageProviderTest {
...
+        verify(exactly = 1) { callback.onFailure(any()) }
+        assertEquals(CLIENT_ERROR, messageSlot.captured)
+
```
[the remark this answers is not in the document]
Comment (korshaknn) [reply libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt:103]: done

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt [hunk 11]
```diff
@@ -90,46 +83,30 @@ class GuidanceViewImageProviderTest {
...
-            }
-
-            override fun onFailure(message: String?) {
-                failureRef.set(true)
-                failureMessageRef.set(message)
-                latch.countDown()
-            }
-        }
+    companion object {
+        private const val CLIENT_ERROR = "Client Error"
```
[the remark this answers is not in the document]
Comment (korshaknn) [reply libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt:108]: done

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt [hunk 12]
```diff
@@ -90,46 +83,30 @@ class GuidanceViewImageProviderTest {
...
+        every { callback.onFailure(capture(messageSlot)) } answers { latch.countDown() }
+
+        guidanceViewImageProvider.renderGuidanceView(bannerInstructions, callback)
 
         latch.await()
-        assertTrue(failureRef.get())
-        assertEquals("Client Error", failureMessageRef.get())
+
+        verify(exactly = 1) { callback.onFailure(any()) }
+        assertEquals(CLIENT_ERROR, messageSlot.captured)
```
[the remark this answers is not in the document]
Comment (korshaknn) [reply libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt:102]: done

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt (hunk 1 above)
Comment (korshaknn) [reply libnavigation-ui/src/test/java/com/mapbox/navigation/ui/instruction/GuidanceViewImageProviderTest.kt:80]: done

**Comment by korshaknn** [comment]
@Guardiola31337 ptal

**Comment by abhishek1508** [comment]
Why are we using `@MainCoroutineRule` and `ThreadController` in the tests? We are not using coroutine anywhere in `GuidanceViewImageProvider`?
cc @Guardiola31337 @korshaknn

**Comment by abhishek1508** [comment]
Never mind. I see coroutines got added back to `GuidanceViewImageProvider`

**Comment by Guardiola31337** [comment]
@abhishek1508 Yeah Coroutines approach was reverted in https://github.com/mapbox/mapbox-navigation-android/pull/3405 after finding out was an issue with the MockWebServer setup.

We should be good here 🚀

**Pull request merged**